loudml/prometheus.py

```diff
@@ -0,0 +1,392 @@
+"""
+Prometheus module for Loud ML.
+
+ * PrometheusClient class handles connection to Prometheus
+ * PrometheusBucket class is capable to read and write timed data
+"""
+
+import logging
+import requests
+import json
+
+import numpy as np
+
+from voluptuous import (
+    Required,
+    Optional,
+    Boolean,
+)
+
+from . import (
+    errors,
+    schemas,
+)
+from loudml.misc import (
+    escape_quotes,
+    escape_doublequotes,
+    make_ts,
+    parse_addr,
+    str_to_ts,
+    ts_to_str,
+)
+from loudml.bucket import Bucket
+
+# Prometheus aggregation function:
+# https://prometheus.io/docs/prometheus/latest/querying/operators/#aggregation-operators
+# sum (calculate sum over dimensions)
+# min (select minimum over dimensions)
+# max (select maximum over dimensions)
+# avg (calculate the average over dimensions)
+# stddev (calculate population standard deviation over dimensions)
+# stdvar (calculate population standard variance over dimensions)
+# count (count number of elements in the vector)
+# count_values (count number of elements with the same value)
+# bottomk (smallest k elements by sample value)
```

Comment:
Adding bottomk and topk to AGGREGATORS gives more flexibility to users.

Reply:
agree